libvmi/driver/xen/xen_events.c

```diff
@@ -1159,10 +1134,81 @@ status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
         /*
          * Put the response on the ring
          */
-        put_response(&xe->vm_event, &rsp);
+        put_response(&xe->vm_event, rsp);
         dbprint(VMI_DEBUG_XEN, "--Finished handling event.\n");
     }
 
+    return vrc;
+}
+
+status_t xen_events_listen(vmi_instance_t vmi, uint32_t timeout)
+{
+    vm_event_request_t req;
+    vm_event_response_t rsp;
+    xc_interface * xch = xen_get_xchandle(vmi);
+    xen_events_t * xe = xen_get_events(vmi);
+    domid_t dom = xen_get_domainid(vmi);
+
+    int rc = -1;
+    status_t vrc = VMI_SUCCESS;
+
+    if ( !xch ) {
+        errprint("%s error: invalid xc_interface handle\n", __FUNCTION__);
+        return VMI_FAILURE;
+    }
+    if ( !xe ) {
+        errprint("%s error: invalid xen_events_t handle\n", __FUNCTION__);
+        return VMI_FAILURE;
+    }
+    if ( dom == VMI_INVALID_DOMID ) {
+        errprint("%s error: invalid domid\n", __FUNCTION__);
+        return VMI_FAILURE;
+    }
+
+    // Set whether the access listener is required
+    rc = xc_domain_set_access_required(xch, dom, vmi->event_listener_required);
+    if ( rc < 0 ) {
+        errprint("Error %d setting mem_access listener required to %d\n",
+            rc, vmi->event_listener_required);
+    }
+
+    if(!vmi->shutting_down && timeout > 0) {
+        dbprint(VMI_DEBUG_XEN, "--Waiting for xen events...(%"PRIu32" ms)\n", timeout);
+        rc = wait_for_event_or_timeout(xch, xe->vm_event.xce_handle, timeout);
+        if ( rc < -1 ) {
+            errprint("Error while waiting for event.\n");
+            return VMI_FAILURE;
+        }
+    }
+
+    vrc = process_requests(vmi, &req, &rsp);
+
+    /*
+     * The only way to gracefully handle vmi_clear_event requests
+     * that were issued in a callback is to ensure no more requests
+     * are in the ringpage. We do this by pausing the domain (all vCPUs)
+     * and process all reamining events on the ring. Once no more requests
+     * are on the ring we can remove the events.
+     */
+    if ( g_hash_table_size(vmi->clear_events) ) {
+        vmi_pause_vm(vmi); // Pause all vCPUs
+        vrc = process_requests(vmi, &req, &rsp);
+
+        GHashTableIter i;
+        vmi_event_t **key = NULL;
+        void (*cb)(vmi_event_t *event);
+
+        ghashtable_foreach(vmi->clear_events, i, key, cb) {
+            vmi_clear_event(vmi, *key, cb);
+        }
+
+        g_hash_table_destroy(vmi->clear_events);
+        vmi->clear_events =
+            g_hash_table_new_full(g_int64_hash, g_int64_equal, free, NULL);
+
+        vmi_resume_vm(vmi);
```

Comment:
This actually may be problematic because it will also unpause the vCPU that was paused by the vm_event, before the response was pulled form the ring on the Xen side..